include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -45,37 +47,87 @@ namespace __par_backend_hetero
-    {
-        const _Index __q = __i_elem;                         //diagonal index
-        const _Index __n_diag = std::min<_Index>(__q, __n1); //diagonal size
-        auto __res =
-            std::lower_bound(__diag_it, __diag_it + __n_diag, 1 /*value to find*/,
-                             [&__rng2, &__rng1, __q, __comp](const auto& __i_diag, const auto& __value) mutable {
-                                 const auto __zero_or_one = __comp(__rng2[__q - __i_diag - 1], __rng1[__i_diag]);
-                                 return __zero_or_one < __value;
-                             });
-        return std::make_pair(*__res, __q - *__res);
-    }
-    else
-    {
-        const _Index __q = __i_elem - __n2;                         //diagonal index
-        const _Index __n_diag = std::min<_Index>(__n1 - __q, __n2); //diagonal size
-        auto __res =
-            std::lower_bound(__diag_it, __diag_it + __n_diag, 1 /*value to find*/,
-                             [&__rng2, &__rng1, __n2, __q, __comp](const auto& __i_diag, const auto& __value) mutable {
-                                 const auto __zero_or_one = __comp(__rng2[__n2 - __i_diag - 1], __rng1[__q + __i_diag]);
-                                 return __zero_or_one < __value;
-                             });
-        return std::make_pair(__q + *__res, __n2 - *__res);
-    }
+    using _IndexSigned = std::make_signed_t<_Index>;
+
+    ////////////////////////////////////////////////////////////////////////////////////
+    // Taking into account the specified constraints of the range of processed data
+    const _IndexSigned __index_sum = __i_elem - 1;
+
+    _IndexSigned idx1_from = __rng1_from;
+    _IndexSigned idx1_to = __rng1_to;
+
+    _IndexSigned idx2_from = __index_sum - (__rng1_to - 1);
+    _IndexSigned idx2_to = __index_sum - __rng1_from + 1;
+
+    const _IndexSigned idx2_from_diff =
+        idx2_from < (_IndexSigned)__rng2_from ? (_IndexSigned)__rng2_from - idx2_from : 0;
+    const _IndexSigned idx2_to_diff = idx2_to > (_IndexSigned)__rng2_to ? idx2_to - (_IndexSigned)__rng2_to : 0;
+
+    idx1_to -= idx2_from_diff;
+    idx1_from += idx2_to_diff;
+
+    idx2_from = __index_sum - (idx1_to - 1);
+    idx2_to = __index_sum - idx1_from + 1;
+
+    ////////////////////////////////////////////////////////////////////////////////////
+    // Run search of split point on diagonal
+
+    using __it_t = oneapi::dpl::counting_iterator<_Index>;
+
+    __it_t __diag_it_begin(idx1_from);
+    __it_t __diag_it_end(idx1_to);
+
+    constexpr bool kValue = false;
+    const __it_t __res = std::lower_bound(__diag_it_begin, __diag_it_end, kValue,
+                                          [&__rng1, &__rng2, __index_sum, __comp](_Index __idx, const bool __value) mutable {
+                                              return __value == __comp(__rng2[__index_sum - __idx], __rng1[__idx]);
```

Comment:
It's enough here to compare the result of `__comp` call with the `__value` for equality.